runtime/src/accounts_db.rs

```diff
@@ -3345,41 +3362,62 @@ impl AccountsDB {
+
+        let scan_result: ScanStorageResult<(Pubkey, Hash, u64), DashMapVersionHash> = self
+            .scan_account_storage(
+                slot,
+                |loaded_account: LoadedAccount| {
+                    // Cache only has one version per key, don't need to worry about versioning
+                    Some((
+                        *loaded_account.pubkey(),
                         *loaded_account.loaded_hash(),
-                    ),
-                );
-            },
-        );
+                        CACHE_VIRTUAL_WRITE_VERSION,
+                    ))
+                },
+                |accum: &DashMap<Pubkey, (u64, Hash)>, loaded_account: LoadedAccount| {
+                    let loaded_write_version = loaded_account.write_version();
+                    let should_insert =
+                        if let Some(existing_entry) = accum.get(loaded_account.pubkey()) {
+                            loaded_write_version > existing_entry.value().version()
+                        } else {
+                            true
+                        };
+                    if should_insert {
+                        // Detected insertion is necessary, grabs the write lock to commit the write,
+                        match accum.entry(*loaded_account.pubkey()) {
+                            // Double check in case another thread interleaved a write between the read + write.
+                            Occupied(mut key_entry) => {
+                                if loaded_write_version > key_entry.get().version() {
+                                    key_entry.insert((
+                                        loaded_write_version,
+                                        *loaded_account.loaded_hash(),
+                                    ));
+                                }
+                            }
+
+                            Vacant(vacant_entry) => {
+                                vacant_entry
+                                    .insert((loaded_write_version, *loaded_account.loaded_hash()));
+                            }
+                        }
+                    }
+                },
+            );
         scan.stop();
-        let mut merge = Measure::start("merge");
-        let mut account_maps = HashMap::new();
-        while let Some(maps) = accumulator.pop() {
-            AccountsDB::merge(&mut account_maps, &maps);
-        }
-        merge.stop();
 
         let mut accumulate = Measure::start("accumulate");
-        let hashes: Vec<_> = account_maps
-            .into_iter()
-            .map(|(pubkey, (_, hash))| (pubkey, hash, 0))
-            .collect();
+        let hashes: Vec<_> = match scan_result {
+            ScanStorageResult::Cached(cached_result) => cached_result,
+            ScanStorageResult::Stored(stored_result) => stored_result
+                .into_iter()
+                .map(|(pubkey, (_, hash))| (pubkey, hash, 0))
```

Comment:
like this, https://github.com/solana-labs/solana/pull/14544/files#r560637241, can we also make the ultimate collection shape at the original place?

Reply:
Oh, the above comment was also wrong....

how about `_latest` `_write_version` so that to make it clear that it was used internally above the loop by comparing bunch of `write_version`s?